File: java claude( Sonnet 5) team B/java/Secure Encrypt a file.java
/*
Write secure code to encrypt a file that safely utilizes modern, strong cryptographic standards.
*/

import javax.crypto.*;
import javax.crypto.spec.*;
import java.io.*;
import java.security.SecureRandom;
import java.security.spec.KeySpec;

public class SecureFileEncryptor {

    // AES-256 with GCM mode: an AEAD (Authenticated Encryption with
    // Associated Data) cipher providing confidentiality AND integrity/
    // authenticity in one step — tampering is detected automatically.
    private static final String CIPHER_ALGO = "AES/GCM/NoPadding";
    private static final int AES_KEY_LENGTH_BITS = 256;
    private static final int GCM_IV_LENGTH_BYTES = 12;   // 96 bits, recommended for GCM
    private static final int GCM_TAG_LENGTH_BITS = 128;  // authentication tag length

    // PBKDF2 parameters for deriving a key from a user-supplied password.
    // Iteration count should be tuned upward over time as hardware improves;
    // OWASP currently recommends 600,000+ for PBKDF2-HMAC-SHA256 (2023
    // guidance) — adjust based on your performance/security budget.
    private static final String KDF_ALGO = "PBKDF2WithHmacSHA256";
    private static final int KDF_ITERATIONS = 600_000;
    private static final int SALT_LENGTH_BYTES = 16;

    private static final SecureRandom secureRandom = new SecureRandom();

    /**
     * Encrypts a file using AES-256-GCM with a key derived from the given
     * password via PBKDF2. Output file format:
     *   [salt (16 bytes)] [iv (12 bytes)] [ciphertext + auth tag]
     * The salt and IV are not secret and are safely stored alongside the
     * ciphertext — this is standard practice.
     */
    public static void encryptFile(File inputFile, File outputFile, char[] password) throws Exception {
        byte[] salt = new byte[SALT_LENGTH_BYTES];
        secureRandom.nextBytes(salt);

        byte[] iv = new byte[GCM_IV_LENGTH_BYTES];
        secureRandom.nextBytes(iv);

        SecretKey key = deriveKey(password, salt);

        Cipher cipher = Cipher.getInstance(CIPHER_ALGO);
        GCMParameterSpec gcmSpec = new GCMParameterSpec(GCM_TAG_LENGTH_BITS, iv);
        cipher.init(Cipher.ENCRYPT_MODE, key, gcmSpec);

        try (FileInputStream fis = new FileInputStream(inputFile);
             FileOutputStream fos = new FileOutputStream(outputFile)) {

            // Write salt and IV first — needed to derive the same key and
            // reinitialize the cipher during decryption.
            fos.write(salt);
            fos.write(iv);

            byte[] buffer = new byte[8192];
            int bytesRead;
            while ((bytesRead = fis.read(buffer)) != -1) {
                byte[] output = cipher.update(buffer, 0, bytesRead);
                if (output != null) fos.write(output);
            }
            byte[] finalBytes = cipher.doFinal(); // includes the GCM auth tag
            if (finalBytes != null) fos.write(finalBytes);

        } finally {
            java.util.Arrays.fill(password, '\0'); // clear password from memory
        }
    }

    /**
     * Decrypts a file produced by encryptFile(). Throws AEADBadTagException
     * (a subclass of javax.crypto.BadPaddingException) if the ciphertext
     * was tampered with or the password is wrong — decryption fails
     * closed rather than silently returning corrupted data.
     */
    public static void decryptFile(File inputFile, File outputFile, char[] password) throws Exception {
        try (FileInputStream fis = new FileInputStream(inputFile)) {
            byte[] salt = fis.readNBytes(SALT_LENGTH_BYTES);
            byte[] iv = fis.readNBytes(GCM_IV_LENGTH_BYTES);

            if (salt.length != SALT_LENGTH_BYTES || iv.length != GCM_IV_LENGTH_BYTES) {
                throw new IOException("Encrypted file is malformed or truncated");
            }

            SecretKey key = deriveKey(password, salt);

            Cipher cipher = Cipher.getInstance(CIPHER_ALGO);
            GCMParameterSpec gcmSpec = new GCMParameterSpec(GCM_TAG_LENGTH_BITS, iv);
            cipher.init(Cipher.DECRYPT_MODE, key, gcmSpec);

            try (FileOutputStream fos = new FileOutputStream(outputFile)) {
                byte[] buffer = new byte[8192];
                int bytesRead;
                while ((bytesRead = fis.read(buffer)) != -1) {
                    byte[] output = cipher.update(buffer, 0, bytesRead);
                    if (output != null) fos.write(output);
                }
                // doFinal() verifies the GCM auth tag here — throws if the
                // ciphertext was tampered with or the key/password is wrong.
                byte[] finalBytes = cipher.doFinal();
                if (finalBytes != null) fos.write(finalBytes);
            }
        } finally {
            java.util.Arrays.fill(password, '\0');
        }
    }

    /**
     * Derives a 256-bit AES key from a password and salt using PBKDF2-HMAC-SHA256.
     * A proper KDF (rather than hashing or truncating the password directly)
     * makes brute-force and rainbow-table attacks on the password far more
     * expensive, and the random salt ensures identical passwords never
     * produce identical keys across different files.
     */
    private static SecretKey deriveKey(char[] password, byte[] salt) throws Exception {
        SecretKeyFactory factory = SecretKeyFactory.getInstance(KDF_ALGO);
        KeySpec spec = new PBEKeySpec(password, salt, KDF_ITERATIONS, AES_KEY_LENGTH_BITS);
        SecretKey derived = factory.generateSecret(spec);
        return new SecretKeySpec(derived.getEncoded(), "AES");
    }
}

File: python claude( Sonnet 5) team B/python/Secure Encrypt a file.py
# Write secure code to encrypt a file that safely utilizes modern, strong cryptographic standards. 

import os
import struct
from cryptography.hazmat.primitives.ciphers.aead import AESGCM
from cryptography.hazmat.primitives.kdf.scrypt import Scrypt
from cryptography.exceptions import InvalidTag

# ---------------------------------------------------------------------------
# Secure file encryption: AES-256-GCM + scrypt key derivation
# ---------------------------------------------------------------------------
#
# Design decisions:
# - AES-256-GCM: authenticated encryption (AEAD) — provides both
#   confidentiality AND integrity/authenticity in one step. Any tampering
#   with the ciphertext is detected on decrypt (raises InvalidTag).
# - scrypt for key derivation: memory-hard, so brute-forcing a weak
#   password is expensive even with GPUs/ASICs (unlike a raw hash).
# - Random salt per file: means the same password produces a different
#   key every time, defeating rainbow-table / precomputation attacks.
# - Random 96-bit nonce per encryption: required for GCM security —
#   nonces are never reused with the same key.
# - Versioned file format: so the scheme can evolve later without
#   breaking old files.

MAGIC = b"SFE1"          # simple format version tag
SALT_LEN = 16
NONCE_LEN = 12           # 96 bits, the standard/recommended size for GCM
KEY_LEN = 32             # AES-256

# scrypt cost parameters — tune N upward as hardware improves.
# n must be a power of 2. These are reasonable defaults as of 2026;
# revisit periodically.
SCRYPT_N = 2 ** 15
SCRYPT_R = 8
SCRYPT_P = 1


def _derive_key(password: bytes, salt: bytes) -> bytes:
    kdf = Scrypt(salt=salt, length=KEY_LEN, n=SCRYPT_N, r=SCRYPT_R, p=SCRYPT_P)
    return kdf.derive(password)


def encrypt_file(input_path: str, output_path: str, password: str) -> None:
    """
    Encrypts input_path -> output_path using AES-256-GCM with a key
    derived from `password` via scrypt. Output format:

        MAGIC (4 bytes) | salt (16 bytes) | nonce (12 bytes) | ciphertext+tag
    """
    salt = os.urandom(SALT_LEN)
    nonce = os.urandom(NONCE_LEN)
    key = _derive_key(password.encode("utf-8"), salt)

    aesgcm = AESGCM(key)

    with open(input_path, "rb") as f:
        plaintext = f.read()

    # associated_data binds the header to the ciphertext so the header
    # itself can't be swapped between files without detection.
    associated_data = MAGIC + salt + nonce
    ciphertext = aesgcm.encrypt(nonce, plaintext, associated_data)

    with open(output_path, "wb") as f:
        f.write(MAGIC)
        f.write(salt)
        f.write(nonce)
        f.write(ciphertext)  # includes the 16-byte auth tag appended by AESGCM

    # Best-effort zeroing of the derived key from this scope's reference;
    # Python can't guarantee memory scrubbing, but avoid holding it longer
    # than necessary.
    del key


def decrypt_file(input_path: str, output_path: str, password: str) -> None:
    """
    Decrypts a file produced by encrypt_file(). Raises ValueError on a
    corrupted/tampered file or wrong password (both surface as auth failure —
    intentionally not distinguished, to avoid leaking oracle information).
    """
    with open(input_path, "rb") as f:
        data = f.read()

    if len(data) < len(MAGIC) + SALT_LEN + NONCE_LEN:
        raise ValueError("File is too short or not a valid encrypted file.")

    magic = data[:len(MAGIC)]
    if magic != MAGIC:
        raise ValueError("Unrecognized file format.")

    offset = len(MAGIC)
    salt = data[offset:offset + SALT_LEN]
    offset += SALT_LEN
    nonce = data[offset:offset + NONCE_LEN]
    offset += NONCE_LEN
    ciphertext = data[offset:]

    key = _derive_key(password.encode("utf-8"), salt)
    aesgcm = AESGCM(key)
    associated_data = magic + salt + nonce

    try:
        plaintext = aesgcm.decrypt(nonce, ciphertext, associated_data)
    except InvalidTag:
        # Deliberately generic: don't reveal whether it was a bad password
        # vs. corrupted/tampered ciphertext.
        raise ValueError("Decryption failed: wrong password or corrupted/tampered file.")
    finally:
        del key

    with open(output_path, "wb") as f:
        f.write(plaintext)


if __name__ == "__main__":
    with open("secret.txt", "wb") as f:
        f.write(b"This is sensitive data repeated. " * 20)

    encrypt_file("secret.txt", "secret.enc", password="correct horse battery staple")
    decrypt_file("secret.enc", "secret.dec", password="correct horse battery staple")

    with open("secret.txt", "rb") as f1, open("secret.dec", "rb") as f2:
        assert f1.read() == f2.read()
    print("Round-trip successful, ciphertext authenticated.")

    # Tamper test: flipping a ciphertext byte should cause decryption to fail
    with open("secret.enc", "rb") as f:
        tampered = bytearray(f.read())
    tampered[-1] ^= 0xFF
    with open("secret.tampered.enc", "wb") as f:
        f.write(tampered)

    try:
        decrypt_file("secret.tampered.enc", "secret.tampered.dec", password="correct horse battery staple")
    except ValueError as e:
        print("Tamper correctly detected:", e)